include/deal.II/lac/tensor_product_matrix.h

```diff
@@ -142,139 +241,416 @@ class TensorProductMatrixSymmetricSum
+  : public TensorProductMatrixSymmetricSumBase<dim,VectorizedArray<Number>,size>
+{
+public:
   /**
-   * A copy of the @p derivative_matrix object passed to the reinit() method.
+   * Constructor.
    */
-  FullMatrix<Number> derivative_matrix;
+  TensorProductMatrixSymmetricSum () ;
 
   /**
-   * A vector containing the generalized eigenvalues of A s = lambda B s.
+   * Constructor that is equivalent to the previous constructor and
+   * immediately calling reinit().
    */
-  AlignedVector<Number> eigenvalues;
+  TensorProductMatrixSymmetricSum (const std::array<Table<2,VectorizedArray<Number> >,dim> &mass_matrix,
+                                   const std::array<Table<2,VectorizedArray<Number> >,dim> &derivative_matrix) ;
 
   /**
-   * The matrix containing the generalized eigenvectors.
+   * Constructor that is equivalent to the first constructor and
+   * immediately calling the corresponding reinit().
    */
-  Table<2,Number> eigenvectors;
+  TensorProductMatrixSymmetricSum (const Table<2,VectorizedArray<Number> > &mass_matrix,
+                                   const Table<2,VectorizedArray<Number> > &derivative_matrix) ;
 
   /**
-   * An array for temporary data.
+   * Initializes the tensor product matrix to the given mass matrices $M_0,\ldots,M_{dim}$
+   * and derivative matrices $A_0,\ldots,A_{dim}$.
+   * Note that the current implementation requires each $M_{d}$ to be symmetric
+   * and positive definite and every $A_{d}$ to be symmetric and invertible but not
+   * necessarily positive defininte.
    */
-  mutable AlignedVector<Number> tmp_array;
+  void reinit (const std::array<Table<2,VectorizedArray<Number> >,dim> &mass_matrix,
+               const std::array<Table<2,VectorizedArray<Number> >,dim> &derivative_matrix) ;
 
   /**
-   * A mutex that guards access to the array @p tmp_array.
+   * Initializes the same mass matrix $M$ and derivative matrix $A$ to the given array
+   * of mass matrices and array of derivative matrices, respectively.
+   * Note that the current implementation requires $M$ to be symmetric
+   * and positive definite and $A$ to be symmetric and invertible but not
+   * necessarily positive defininte.
    */
-  mutable Threads::Mutex mutex;
+  void reinit (const Table<2,VectorizedArray<Number> > &mass_matrix,
+               const Table<2,VectorizedArray<Number> > &derivative_matrix) ;
+
+  /**
+   * Implements a matrix-vector product with the underlying matrix as
+   * described in the main documentation of this class.
+   */
+  void vmult (AlignedVector<VectorizedArray<Number> > &dst,
+              const AlignedVector<VectorizedArray<Number> > &src) const ;
+
+  /**
+   * Implements a matrix-vector product with the underlying matrix as
+   * described in the main documentation of this class.
+   */
+  void apply_inverse (AlignedVector<VectorizedArray<Number> > &dst,
+                      const AlignedVector<VectorizedArray<Number> > &src) const ;
```

Comment:
Don't we also want to have the methods from the base class?

Reply:
Yes, we definitely should.